### PR TITLE
fix(core): Implement stricter variable typings on generic

### LINKED ISSUE #2606: RFC: Fix for useQuery variables being optional
In my codebase this feels like a bug - but I'm posting it as an RFC since I might be missing some perspective on this. An example of what I expect to give a type error:
```GraphQL
query GetPage($pageNumber: Int!) {
  page(pageNumber: $pageNumber){
    title
    body
  }
}
```

Which doesn't give a type error for this:
```TypeScript
const [{ data }] = useGetPage() // <-- obviously missing variables with pageNumber
```

When using useQuery the variables parameter is always optional - even if variables are required by the query type. The issue is with the `UseQueryArgs` interface, which currently looks like this:
```TypeScript
export interface UseQueryArgs<Variables = object, Data = any> {
    query: string | DocumentNode | TypedDocumentNode<Data, Variables>;
    variables?: Variables; // <--- here is the issue
    requestPolicy?: RequestPolicy;
    context?: Partial<OperationContext>;
    pause?: boolean;
}
```

I did some experimenting and I think something like this could be better (works as expected when using useQuery, although it does cause some type compatibility issues with [@graphql-codegen/typescript-urql](https://npmjs.com/package/@graphql-codegen/typescript-urql)):

```TypeScript
export interface UseQueryArgsWithVariables<Variables = object, Data = any> {
    query: string | DocumentNode | TypedDocumentNode<Data, Variables>;
    variables: Variables;
    requestPolicy?: RequestPolicy;
    context?: Partial<OperationContext>;
    pause?: boolean;
}

export interface UseQueryArgsWithoutVariables<Data = any> {
    query: string | DocumentNode | TypedDocumentNode<Data, Variables>;
    requestPolicy?: RequestPolicy;
    context?: Partial<OperationContext>;
    pause?: boolean;
}

export type UseQueryArgs<Variables = object, Data = any> = keyof Variables extends never ? UseQueryArgsWithoutVariables<Data> : UseQueryArgsWithVariables<Variables, Data>
```

I tried with a similar approach for useMutation which has the same issue but wasn't able to get it 100%.

### PR DESCRIPTION
**Pull request by kitten**
Resolve #2606

## Summary

This attempts to introduce a stricter type for variables preventing it from being an optional property or argument (i.e. `variables?: Variables`) in all parts of the codebase. To do this, we also need to make its type a little more strict with `AnyVariables`, which explicitly defaults the generic to a type of `void | undefined | { ... }`. In case of a generic that's passed in, this also becomes the extension type.

This is currently a draft because some methods on `Client` and some utility functions, like `createRequest`, are affected by this change. We have to check that these behave as expected (as per the change and RFC) and that we don't break them if that isn't necessary

## Set of changes

- Add `AnyVariables` and make generics `Variables extends AnyVariables = AnyVariables`
- Update core typings where `variables` were optional
- Update all typings across bindings (wherever `variables?:` was used, a special type must be implemented)

### CODE REVIEW

**Review by JoviDeCroock — APPROVED** [review]

File: packages/vue-urql/src/useQuery.ts [hunk 1]
```diff
@@ -23,19 +24,25 @@ import { unwrapPossibleProxy } from './utils';
...
-}
-
-export type QueryPartialState<T = any, V = object> = Partial<
-  OperationResult<T, V>
-> & { fetching?: boolean };
-
-export interface UseQueryState<T = any, V = object> {
+} & (V extends void
+  ? {
+      variables?: MaybeRef<V>;
```
Comment (JoviDeCroock) [inline packages/vue-urql/src/useQuery.ts:34]:
```suggestion
      variables?: MaybeRef<{ [K in keyof V]: MaybeRef<V[K]> }>;
```

File: packages/vue-urql/src/useQuery.ts [hunk 2]
```diff
@@ -23,19 +24,25 @@ import { unwrapPossibleProxy } from './utils';
...
+    }
+  : {
+      variables: MaybeRef<V>;
```
Comment (JoviDeCroock) [inline packages/vue-urql/src/useQuery.ts:37]:
```suggestion
      variables: MaybeRef<{ [K in keyof V]: MaybeRef<V[K]> }>;;
```

File: packages/vue-urql/src/useSubscription.ts [hunk 3]
```diff
@@ -21,17 +22,29 @@ import { unwrapPossibleProxy } from './utils';
...
+  V extends AnyVariables = AnyVariables
+> = {
   query: MaybeRef<TypedDocumentNode<T, V> | DocumentNode | string>;
-  variables?: MaybeRef<V>;
   pause?: MaybeRef<boolean>;
   context?: MaybeRef<Partial<OperationContext>>;
-}
+} & (V extends void
+  ? {
+      variables?: MaybeRef<V>;
```
Comment (JoviDeCroock) [inline packages/vue-urql/src/useSubscription.ts:34]:
```suggestion
      variables?: MaybeRef<{ [K in keyof V]: MaybeRef<V[K]> }>;
```

File: packages/vue-urql/src/useSubscription.ts [hunk 4]
```diff
@@ -21,17 +22,29 @@ import { unwrapPossibleProxy } from './utils';
...
+    }
+  : {
+      variables: MaybeRef<V>;
```
Comment (JoviDeCroock) [inline packages/vue-urql/src/useSubscription.ts:37]:
```suggestion
      variables: MaybeRef<{ [K in keyof V]: MaybeRef<V[K]> }>;
```

**Review by JoviDeCroock — COMMENTED** [review]
These come from #2608

**Comment by crmitchelmore** [comment]
@kitten is the expected pattern to pass empty variables for cases when there are no variables?

e.g. `client.query(SomeDocument)`

becomes:

`client.query(SomeDocument, {})`

**Pull request merged**